src/test/FuzzerImpl.cpp

```diff
@@ -0,0 +1,358 @@
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "test/FuzzerImpl.h"
+#include "ledger/LedgerTxn.h"
+#include "main/Application.h"
+#include "main/Config.h"
+#include "overlay/OverlayManager.h"
+#include "simulation/Simulation.h"
+#include "test/TestUtils.h"
+#include "test/test.h"
+#include "transactions/OperationFrame.h"
+#include "transactions/SignatureChecker.h"
+#include "util/Math.h"
+
+#include <xdrpp/autocheck.h>
+
+namespace stellar
+{
+
+// creates a generic configuration with settings rigged to maximize
+// determinism
+static Config
+getFuzzConfig(int instanceNumber)
+{
+    Config cfg = getTestConfig(instanceNumber);
+    cfg.MANUAL_CLOSE = true;
+    cfg.CATCHUP_COMPLETE = false;
+    cfg.CATCHUP_RECENT = 0;
+    cfg.ARTIFICIALLY_GENERATE_LOAD_FOR_TESTING = false;
+    cfg.ARTIFICIALLY_SET_CLOSE_TIME_FOR_TESTING = UINT32_MAX;
+    cfg.PUBLIC_HTTP_PORT = false;
+    cfg.WORKER_THREADS = 1;
+    cfg.QUORUM_INTERSECTION_CHECKER = false;
+    cfg.PREFERRED_PEERS_ONLY = false;
+    cfg.RUN_STANDALONE = true;
+
+    return cfg;
+}
+
+void
+seedRandomness()
+{
+    srand(1);
+    gRandomEngine.seed(1);
+}
+
+// FuzzTransactionFrame is a specialized TransactionFrame that includes
+// useful methods for fuzzing such as an attemptApplication method for reseting
+// ledger state and deterministically attempting application of transactions.
+class FuzzTransactionFrame : public TransactionFrame
+{
+  public:
+    FuzzTransactionFrame(Hash const& networkID,
+                         TransactionEnvelope const& envelope)
+        : TransactionFrame(networkID, envelope){};
+
+    void
+    attemptApplication(Application& app, AbstractLedgerTxn& ltx)
+    {
+        // seed randomness
+        seedRandomness();
+
+        // before application, clear cache so that they don't persist
+        // between executions and confuse the fuzzer, causing it to
+        // think some Tx with [op0...opN] can result in multiple
+        // different execution paths.
+        app.getLedgerTxnRoot().resetForFuzzer();
+
+        // reset results of operations
+        resetResults(ltx.getHeader(), 0);
+
+        // attempt application of transaction without accounting for sequence
+        // number, processing the fee, or committing the LedgerTxn
+        mCachedAccount.reset();
```

Comment:
why do you need to reset `mCachedAccount` ? Don't we want to let it do its thing depending on protocol version?

Reply:
This was part of the [`TransactionFrame::apply` method](https://github.com/stellar/stellar-core/blob/master/src/transactions/TransactionFrame.cpp#L655) I recreated here. I believe this reset for `mCachedAccount` seems to happen regardless of protocol.

Reply:
I asked because `loadSourceAccount` seems to already do the right thing

Reply:
Yes, good point -- this was for a large portion of the fee processing and signature checking I left out. Removed.